File: python/rafter_cli/core/docs_loader.py
"""Resolve, fetch, and cache repo-specific security docs from .rafter.yml."""
from __future__ import annotations

import hashlib
import json
import os
import sys
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable
from urllib.error import URLError
from urllib.request import Request, urlopen

from ..utils.git import get_git_root
from .config_schema import get_rafter_dir
from .policy_loader import load_policy


DEFAULT_TTL_SECONDS = 86400
_FETCH_TIMEOUT = 15


@dataclass
class ResolvedDoc:
    id: str
    source: str
    source_kind: str  # "path" | "url"
    description: str
    tags: list[str]
    cache_status: str  # "local" | "cached" | "not-cached" | "stale"
    ttl_seconds: int | None = None


@dataclass
class FetchResult:
    content: str
    cached: bool
    stale: bool
    source: str
    source_kind: str


def _cache_dir() -> Path:
    return get_rafter_dir() / "docs-cache"


def _cache_key(url: str) -> str:
    return hashlib.sha256(url.encode("utf-8")).hexdigest()[:32]


def _cache_paths(url: str) -> tuple[Path, Path]:
    base = _cache_dir() / _cache_key(url)
    return base.with_suffix(".txt"), base.with_suffix(".meta.json")


def _read_cache(url: str) -> tuple[str, float] | None:
    content_path, meta_path = _cache_paths(url)
    if not content_path.exists() or not meta_path.exists():
        return None
    try:
        meta = json.loads(meta_path.read_text())
        body = content_path.read_text()
        fetched_at_str = meta.get("fetched_at", "")
        # Parse ISO timestamp (simple fallback using fromisoformat)
        from datetime import datetime
        dt = datetime.fromisoformat(fetched_at_str.replace("Z", "+00:00"))
        return body, dt.timestamp()
    except (OSError, ValueError, json.JSONDecodeError):
        return None


def _write_cache(url: str, body: str, content_type: str) -> None:
    from datetime import datetime, timezone
    cache_dir = _cache_dir()
    cache_dir.mkdir(parents=True, exist_ok=True)
    content_path, meta_path = _cache_paths(url)
    content_path.write_text(body)
    meta_path.write_text(json.dumps({
        "fetched_at": datetime.now(timezone.utc).isoformat(),
        "url": url,
        "content_type": content_type,
    }, indent=2) + "\n")


def _is_expired(fetched_at: float, ttl_seconds: int) -> bool:
    return (time.time() - fetched_at) > ttl_seconds


def _resolve_policy_path(relative: str) -> Path:
    p = Path(relative)
    if p.is_absolute():
        return p
    root = get_git_root() or os.getcwd()
    return Path(root) / p


def list_docs(entries: Iterable[dict] | None = None) -> list[ResolvedDoc]:
    """List docs from the active policy with resolution metadata. No network I/O."""
    if entries is None:
        policy = load_policy() or {}
        entries = policy.get("docs") or []

    resolved: list[ResolvedDoc] = []
    for entry in entries:
        doc_id = entry["id"]
        description = entry.get("description", "")
        tags = list(entry.get("tags") or [])
        if "path" in entry:
            resolved.append(ResolvedDoc(
                id=doc_id,
                source=entry["path"],
                source_kind="path",
                description=description,
                tags=tags,
                cache_status="local",
            ))
            continue
        url = entry["url"]
        ttl = (entry.get("cache") or {}).get("ttl_seconds", DEFAULT_TTL_SECONDS)
        cached = _read_cache(url)
        if cached is None:
            status = "not-cached"
        else:
            status = "stale" if _is_expired(cached[1], ttl) else "cached"
        resolved.append(ResolvedDoc(
            id=doc_id,
            source=url,
            source_kind="url",
            description=description,
            tags=tags,
            cache_status=status,
            ttl_seconds=ttl,
        ))
    return resolved


def resolve_doc_selector(selector: str, entries: Iterable[dict] | None = None) -> list[dict]:
    """Resolve docs matching an id (exact) or tag (any). Returns raw policy entries."""
    if entries is None:
        policy = load_policy() or {}
        entries = list(policy.get("docs") or [])
    else:
        entries = list(entries)

    for entry in entries:
        if entry.get("id") == selector:
            return [entry]
    return [e for e in entries if selector in (e.get("tags") or [])]


def fetch_doc(entry: dict, refresh: bool = False) -> FetchResult:
    """Return content for a doc entry, fetching URL docs on miss/expired/refresh."""
    if "path" in entry:
        path = _resolve_policy_path(entry["path"])
        content = path.read_text()
        return FetchResult(content=content, cached=False, stale=False,
                           source=entry["path"], source_kind="path")

    url = entry["url"]
    ttl = (entry.get("cache") or {}).get("ttl_seconds", DEFAULT_TTL_SECONDS)
    cached = _read_cache(url)
    fresh = cached is not None and not _is_expired(cached[1], ttl)

    if not refresh and fresh:
        return FetchResult(content=cached[0], cached=True, stale=False,
                           source=url, source_kind="url")

    try:
        req = Request(url, headers={"User-Agent": "rafter-cli"})
        with urlopen(req, timeout=_FETCH_TIMEOUT) as resp:
            body_bytes = resp.read()
            content_type = resp.headers.get("content-type", "text/plain")
        body = body_bytes.decode("utf-8", errors="replace")
        _write_cache(url, body, content_type)
        return FetchResult(content=body, cached=False, stale=False,
                           source=url, source_kind="url")
    except (URLError, TimeoutError, OSError) as exc:
        if cached is not None:
            return FetchResult(content=cached[0], cached=True, stale=True,
                               source=url, source_kind="url")
        raise RuntimeError(f"Failed to fetch {url}: {exc}") from exc

File: node/src/core/docs-loader.ts
import fs from "fs";
import path from "path";
import crypto from "crypto";
import { execSync } from "child_process";
import { loadPolicy, PolicyDocEntry } from "./policy-loader.js";
import { getRafterDir } from "./config-defaults.js";

const DEFAULT_TTL_SECONDS = 86400;

export interface ResolvedDoc extends PolicyDocEntry {
  source: string;
  sourceKind: "path" | "url";
  cacheStatus: "local" | "cached" | "not-cached" | "stale";
  cachedPath?: string;
}

export interface FetchResult {
  content: string;
  cached: boolean;
  stale: boolean;
  source: string;
  sourceKind: "path" | "url";
}

function getCacheDir(): string {
  return path.join(getRafterDir(), "docs-cache");
}

function cacheKey(url: string): string {
  return crypto.createHash("sha256").update(url).digest("hex").slice(0, 32);
}

function cachePaths(url: string): { content: string; meta: string } {
  const dir = getCacheDir();
  const key = cacheKey(url);
  return {
    content: path.join(dir, `${key}.txt`),
    meta: path.join(dir, `${key}.meta.json`),
  };
}

function readCache(url: string): { content: string; fetchedAt: number } | null {
  const { content, meta } = cachePaths(url);
  if (!fs.existsSync(content) || !fs.existsSync(meta)) return null;
  try {
    const metaData = JSON.parse(fs.readFileSync(meta, "utf-8"));
    const body = fs.readFileSync(content, "utf-8");
    const fetchedAt = Date.parse(metaData.fetched_at);
    if (isNaN(fetchedAt)) return null;
    return { content: body, fetchedAt };
  } catch {
    return null;
  }
}

function writeCache(url: string, body: string, contentType: string): void {
  const dir = getCacheDir();
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  const { content, meta } = cachePaths(url);
  fs.writeFileSync(content, body, "utf-8");
  fs.writeFileSync(meta, JSON.stringify({
    fetched_at: new Date().toISOString(),
    url,
    content_type: contentType,
  }, null, 2) + "\n", "utf-8");
}

function isExpired(fetchedAt: number, ttlSeconds: number): boolean {
  return Date.now() - fetchedAt > ttlSeconds * 1000;
}

function resolvePolicyPath(relative: string): string {
  if (path.isAbsolute(relative)) return relative;
  let root: string;
  try {
    root = execSync("git rev-parse --show-toplevel", {
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "ignore"],
    }).trim();
  } catch {
    root = process.cwd();
  }
  return path.resolve(root, relative);
}

/**
 * List docs from the active policy with resolution metadata.
 * Never performs network I/O.
 */
export function listDocs(entries?: PolicyDocEntry[]): ResolvedDoc[] {
  const policy = entries ? { docs: entries } : loadPolicy();
  const docs = policy?.docs || [];

  return docs.map(entry => {
    if (entry.path) {
      return {
        ...entry,
        source: entry.path,
        sourceKind: "path" as const,
        cacheStatus: "local" as const,
      };
    }
    const url = entry.url!;
    const ttl = entry.cache?.ttlSeconds ?? DEFAULT_TTL_SECONDS;
    const cached = readCache(url);
    const { content: cachedPath } = cachePaths(url);
    let cacheStatus: ResolvedDoc["cacheStatus"] = "not-cached";
    if (cached) {
      cacheStatus = isExpired(cached.fetchedAt, ttl) ? "stale" : "cached";
    }
    return {
      ...entry,
      source: url,
      sourceKind: "url" as const,
      cacheStatus,
      cachedPath: cached ? cachedPath : undefined,
    };
  });
}

/**
 * Resolve docs matching an id or tag. Exact id first, then any entry with that tag.
 */
export function resolveDocSelector(selector: string, entries?: PolicyDocEntry[]): PolicyDocEntry[] {
  const policy = entries ? { docs: entries } : loadPolicy();
  const docs = policy?.docs || [];
  const byId = docs.find(d => d.id === selector);
  if (byId) return [byId];
  return docs.filter(d => Array.isArray(d.tags) && d.tags.includes(selector));
}

export interface FetchOptions {
  refresh?: boolean;
}

/**
 * Return content for a doc entry, fetching URL docs on miss/expired/refresh.
 * On network failure with stale cache, returns stale content with stale=true.
 */
export async function fetchDoc(entry: PolicyDocEntry, opts: FetchOptions = {}): Promise<FetchResult> {
  if (entry.path) {
    const abs = resolvePolicyPath(entry.path);
    const content = fs.readFileSync(abs, "utf-8");
    return { content, cached: false, stale: false, source: entry.path, sourceKind: "path" };
  }

  const url = entry.url!;
  const ttl = entry.cache?.ttlSeconds ?? DEFAULT_TTL_SECONDS;
  const cached = readCache(url);
  const fresh = cached && !isExpired(cached.fetchedAt, ttl);

  if (!opts.refresh && fresh) {
    return { content: cached.content, cached: true, stale: false, source: url, sourceKind: "url" };
  }

  try {
    const response = await fetch(url, {
      redirect: "follow",
      signal: AbortSignal.timeout(15000),
    });
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }
    const body = await response.text();
    const contentType = response.headers.get("content-type") || "text/plain";
    writeCache(url, body, contentType);
    return { content: body, cached: false, stale: false, source: url, sourceKind: "url" };
  } catch (err) {
    if (cached) {
      return { content: cached.content, cached: true, stale: true, source: url, sourceKind: "url" };
    }
    throw err;
  }
}
